core/embed/bootloader/main.c

```diff
@@ -256,16 +268,55 @@ int main(void) {
 
   display_reinit();
 
+  const image_header *hdr = NULL;
+  vendor_header vhdr;
+  // detect whether the device contains a valid firmware
+  secbool firmware_present = sectrue;
+
+  if (sectrue != read_vendor_header((const uint8_t *)FIRMWARE_START, &vhdr)) {
+    firmware_present = secfalse;
+  }
+
+  if (sectrue == firmware_present) {
+    firmware_present = check_vendor_header_keys(&vhdr);
+  }
+
+  if (sectrue == firmware_present) {
+    firmware_present = check_vendor_header_lock(&vhdr);
+  }
+
+  if (sectrue == firmware_present) {
+    hdr = read_image_header((const uint8_t *)(FIRMWARE_START + vhdr.hdrlen),
+                            FIRMWARE_IMAGE_MAGIC, FIRMWARE_IMAGE_MAXSIZE);
+    if (hdr != (const image_header *)(FIRMWARE_START + vhdr.hdrlen)) {
+      firmware_present = secfalse;
+    }
+  }
+  if (sectrue == firmware_present) {
+    firmware_present = check_image_model(hdr);
+  }
+  if (sectrue == firmware_present) {
+    firmware_present =
+        check_image_header_sig(hdr, vhdr.vsig_m, vhdr.vsig_n, vhdr.vpub);
+  }
+  if (sectrue == firmware_present) {
+    firmware_present =
+        check_image_contents(hdr, IMAGE_HEADER_SIZE + vhdr.hdrlen,
+                             FIRMWARE_SECTORS, FIRMWARE_SECTORS_COUNT);
+  }
+
 #if defined TREZOR_MODEL_T
```

Comment:
we can probably move the MPU configuration. Otherwise i needed to evaluate firmware presence to show correct "empty logo" color. The boot sequence will be different after latest design changes, but we will still need to know ASAP.

Reply:
moved the MPU: https://github.com/trezor/trezor-firmware/commit/f8791ea30cfce094697c61b12e425254724359b9